File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/LicenseListCallback.java
package in.dragonbra.javasteam.steam.handlers.steamapps.callback;

import in.dragonbra.javasteam.enums.EResult;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver.CMsgClientLicenseList;
import in.dragonbra.javasteam.steam.handlers.steamapps.License;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * This callback is fired during logon, informing the client of it's available licenses.
 */
public class LicenseListCallback extends CallbackMsg {

    private EResult result;

    private List<License> licenseList;

    public LicenseListCallback(CMsgClientLicenseList.Builder msg) {
        result = EResult.from(msg.getEresult());

        List<License> licenses = new ArrayList<>();

        for (CMsgClientLicenseList.License l : msg.getLicensesList()) {
            licenses.add(new License(l));
        }

        licenseList = Collections.unmodifiableList(licenses);
    }

    public EResult getResult() {
        return result;
    }

    public List<License> getLicenseList() {
        return licenseList;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/LicenseListCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamapps.callback

import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.enums.EResult
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver.CMsgClientLicenseList
import `in`.dragonbra.javasteam.steam.handlers.steamapps.License
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg

/**
 * This callback is fired during logon, informing the client of it's available licenses.
 */
class LicenseListCallback(packetMsg: IPacketMsg) : CallbackMsg() {

    /**
     * Gets the result of the message.
     */
    val result: EResult

    /**
     * Gets the license list.
     */
    val licenseList: List<License>

    init {
        val licenseListResp = ClientMsgProtobuf<CMsgClientLicenseList.Builder>(
            CMsgClientLicenseList::class.java,
            packetMsg
        )
        val msg = licenseListResp.body

        result = EResult.from(msg.eresult)

        licenseList = msg.licensesList.map { License(it) }
    }
}
